records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -31,47 +33,99 @@ AWS calculates the bounce rate as the number of hard bounces per the total numbe
+```sql
+with data as
+(
+
+select 
+    service_id, 
+    count(id) as total_emails,
+    count(*) FILTER (WHERE notification_status ='permanent-failure' ) AS hard_bounces
+from 
+    notifications 
+where 
+    created_at BETWEEN NOW() - INTERVAL '24 HOURS' AND NOW() -- this value is the `[bounce-rate-window]`
+group by 
+    service_id
+having
+    count(id) >= 1000 -- this value is the `[bounce-rate-warning-notification-volume-minimum]`
+)
+
+select *, 100*(hard_bounces::decimal/total_emails::decimal) as bounce_rate from data
+```
 
 ## Additional considerations
 
-### How GC Notify handles bounce notifications from SES
+### Staggered roll-out
+To prevent suspending services that may not actually be causing issues with the overall bounce rate, it is suggested to roll out this feature in stages.  The first stage would include tracking bounces, calculating bounce rates, and displaying them on service dashboards. It would also include providing bounce rate guidance for services. Once the calculations have been confirmed to work satisfactorily with live data, the second stage can be rolled out which will include warning and suspending services.
+
+### How GC Notify handles asynchronous bounce notifications from SES
 GC Notify uses a lambda function named `ses-to-sqs-email-callbacks` to handle bounce notifications sent back from SES.  When a callback comes in, the lambda function parses the bounce notification and sends it a queue to be picked up and run by the celery task `process-ses-result`.  The `process-ses-result` task updates the status of the notification in the database including whether it was a hard or a soft bounce.
 
+### How GC Notify handles synchronous bounce notifications from SES
+GC Notify sends emails via the `send_email()` method in `clients/email/aws_ses.py`.  
+- :warning: There does not appear to be code to record synchronous bounces in this method.
+
+## Proposed values
+
+The values below can be used as a starting point for this feature.  They should be reviewed and adjusted as needed as the feature is tested and rolled out.
+
+| Parameter                                       | Value    | Notes                                                                                                                        |
+|-------------------------------------------------|----------|------------------------------------------------------------------------------------------------------------------------------|
+| bounce-rate-warning-percentage                  |       5% | AWS indicates a 5% bounce rate as under review, so we should warn no higher than this                                        |
+| bounce-rate-critical-percentage                 |       9% | AWS indicates a 10% or higher bounce rate as at risk, so we should suspend before this                                       |
```

Comment:
We can have this at 10% I think.

Reply:
I was thinking that 10% and greater was a very-bad-thing:tm:, but I will take your recommendation on this :)

Reply:
My thinking is that one service bounce rate of 10% will not get past the global bounce rate past 10% because the other services will amortize the latter. It could be incorrect if only one service sends notifications for a long enough period of time which is unlikely, especially considering that GCNotify constantly sends notifications for heartbeat (for example).

Reply:
Yes this is a good line of thinking and a good assumption!

Reply:
My last comment made me thinking though: would it be possible for GCNotify to get past 10% bounce rate? Could there be a possibility where we'd suspend our own service? We might want to take explicit conditions to take out GCNotify from such automatic suspension and rather raise critical alarms if that every happened. Let's add this to the ADR please. 🙏